python_src/src/lamp_py/migrations/versions/performance_manager/695ef31623ae_rt_trunk_id_and_route_id_processing.py

```diff
@@ -0,0 +1,308 @@
+depends_on = None
+
+
+def upgrade() -> None:
+    op.add_column(
+        "vehicle_trips",
+        sa.Column("branch_route_id", sa.String(60), nullable=True),
+    )
+    create_rt_braintree_function = """
+        CREATE OR REPLACE FUNCTION rt_red_is_braintree_branch(p_trip_hash bytea) RETURNS boolean AS $$ 
+        DECLARE
+            found_check bool;
+        BEGIN 
+            SELECT
+                true
+            INTO
+                found_check
+            FROM 
+                vehicle_events ve
+            WHERE 
+                ve.trip_hash = p_trip_hash
+                AND ve.stop_id IN ('70097', '70098', '70099', '70100', '70101', '70102', '70103', '70104', '70105')
+            LIMIT 1
+            ;
+            IF FOUND THEN
+                RETURN true;
+            ELSE
+                RETURN false;
+            END IF;
+        END;
+        $$ LANGUAGE plpgsql;
+    """
+    op.execute(create_rt_braintree_function)
+
+    create_rt_ashmont_function = """
+        CREATE OR REPLACE FUNCTION rt_red_is_ashmont_branch(p_trip_hash bytea) RETURNS boolean AS $$ 
+        DECLARE
+            found_check bool;
+        BEGIN 
+            SELECT
+                true
+            INTO
+                found_check
+            FROM 
+                vehicle_events ve
+            WHERE 
+                ve.trip_hash = p_trip_hash
+                AND ve.stop_id IN ('70087', '70088', '70089', '70090', '70091', '70092', '70093', '70094')
+            LIMIT 1
+            ;
+            IF FOUND THEN
+                RETURN true;
+            ELSE
+                RETURN false;
+            END IF;
+        END;
+        $$ LANGUAGE plpgsql;
+    """
+    op.execute(create_rt_ashmont_function)
+
+    create_get_rt_branch_trunk_id = """
+        CREATE OR REPLACE FUNCTION update_rt_branch_trunk_id() RETURNS TRIGGER AS $$ 
+        BEGIN 
+            IF NEW.route_id ~ 'Green-' THEN
+                NEW.branch_route_id := NEW.route_id;
+                NEW.trunk_route_id := 'Green';
+            ELSEIF NEW.route_id = 'Red' THEN
+                IF rt_red_is_braintree_branch(NEW.trip_hash) THEN
+                    NEW.branch_route_id := 'Red-Braintree';
+                ELSEIF rt_red_is_ashmont_branch(NEW.trip_hash) THEN
+                    NEW.branch_route_id := 'Red-Ashmont';
+                ELSE
+                    NEW.branch_route_id := null;
+                END IF;
+                NEW.trunk_route_id := NEW.route_id;
+            ELSE
+                NEW.branch_route_id := null;
+                NEW.trunk_route_id := NEW.route_id;
+            END IF;
+            RETURN NEW;
+        END;
+        $$ LANGUAGE plpgsql;
+    """
+    op.execute(create_get_rt_branch_trunk_id)
+
+    create_trigger = """
+        CREATE TRIGGER rt_trips_update_branch_trunk BEFORE INSERT OR UPDATE ON vehicle_trips 
+        FOR EACH ROW EXECUTE PROCEDURE update_rt_branch_trunk_id();
+    """
+    op.execute(create_trigger)
```

Comment:
RT branch / trunk route_id's will be updated by this database trigger. An `INSERT` or `UPDATE` to the trips table will trigger a re-calculation. The `update_trip_stop_counts` function should ensure that any time a trip gets added stop information, theses are re-calculated.